recipes/automated-testing.md

```diff
@@ -86,10 +86,26 @@ changes, we can make minimal changes to the test suite and/or mocked data.
 
 ### Integration Tests
 
-With these tests, we test the application API endpoints and assert that they are
-actually working as expected.
+With these tests, we test the combined behavior of multiple components within the system.
 
-**TODO**: do we want to add that we should run full backend for these type of tests?
+#### Infrastructure
+
+Running the tests on test infrastructure should be preferred to mocking, unlike in unit tests. Ideally, a full application instance would be run, to mimic real application behavior as close as possible.
+This usually includes running the application connected to a test database, inserting fake data into it during the test setup, and doing assertions on the current state of the database. This also means integration test code should have full access to the test infrastructure for querying. 
+> [!NOTE]
+> Regardless of whether using raw queries or the ORM, simple queries should be used to avoid introducing business logic within tests.
+
+However, mocking can still be used when needed, for example when expecting side-effects that call third party services.
+
+#### Entry points
+
+Integration test entry points can vary depending on the application use cases. These include services, controllers, or the API. These are not set in stone and should be taken into account when making a decision. For example:
+- A use case that can be invoked through multiple different protocols can be tested separately from them, to avoid duplication. A tradeoff in this case is the need to write some basic tests for each of the protocols.
+- A use case that will always be invokeable through a single protocol might benefit enough from only being tested using that protocol. E.g. a HTTP API route test might eliminate the need for a lower level, controller/service level test. This would also enable the testing of authorization within the same test, which might not have been possible otherwise depending on the technology used.
```

Comment:
Authorization should be tested separately to increase determinism (if authorization fails, we know it failed immediately).
I assume what you meant here is that this way, we test the integration of the auth layer with other app layers?

Reply:
Yes, was referring to being able to expect a 403 if trying to access an admin route with a non-admin token, for example. Maybe replace `This would also enable the testing of authorization within the same test, which might not have been possible otherwise depending on the technology used.` with `This would also enable the inclusion of the auth layer within these tests, which might not have been possible otherwise depending on the technology used.`?

Reply:
@dbettini That makes more sense to me. Otherwise it might be interpreted as "auth is implicitly tested through other integration tests" so making it clear that this covers testing auth integration makes sense.


> referring to being able to expect a 403 if trying to access an admin route 

Regarding ⬆️ , I guess there are 2 test-cases I see here:
1. Testing is our auth layer working as expected i.e. if for given auth permissions appropriate access is granted
2. Are permissions set properly on specific route/entry point
I think the first should be tested generically in a auth-specific tests while the second we can get implicitly from integration test on that specific entry point. While I see value in testing appropriate permissioning on a entry point, I'm wondering is it worth-while to add (all) auth permutation assertions to each entry point as, depending on the complexity of auth layer and permissioning implemented, it could blow-up number of test-cases. 
What do you think? (I might be looking too much into a simple example 😄 )

Reply:
@kronicker unfortunately, I don't have a catch-all answer there 😄 I do like to test the basic cases, e.g. a route only being accessible to admins, for two reasons: 
1. Other than implicitly testing the behavior of the auth layer, it also tests that the correct rules were applied. This is useful because even if our auth layer works correctly, it still doesn't mean it was correctly configured for this specific route.
2. The level of danger in the case of an error - I feel much safer if a route that e.g. allows destructive actions for admins has a test that ensures other roles would get an error

But like you said, we should make sure to avoid combinatorial explosions, so it needs to be applied on a case by case basis IMO - this is why I didn't go into too much detail in this section, but went deeper in the `When not to use` section.